nat-lab/tests/test_core_api.py

```diff
@@ -0,0 +1,356 @@
+    "ip_addresses": [str],
+    "discovery_key": str,
+    "relay_address": str,
+    "traffic_routing_supported": bool,
+    "is_local": bool,
+    "allow_incoming_connections": bool,
+}
+
+
+derp_structure = {
+    "region_code": str,
+    "name": str,
+    "hostname": str,
+    "ipv4": str,
+    "relay_port": int,
+    "stun_port": int,
+    "stun_plaintext_port": int,
+    "public_key": str,
+    "weight": int,
+}
+
+
+def validate_dict_structure(data_to_validate, expected_data_structure) -> None:
+    for key, expected_type in expected_data_structure.items():
+        assert key in data_to_validate, f"Missing key: {key}"
+        value = data_to_validate[key]
+
+        if isinstance(expected_type, list):
+            assert isinstance(value, list), f"Key '{key}' should be a list."
+            item_type = expected_type[0]
+            for item in value:
+                assert isinstance(
+                    item, item_type
+                ), f"Items in list '{key}' should be of type {item_type}, but got {type(item)}"
+
+        elif isinstance(expected_type, dict):
+            assert isinstance(value, dict), f"Key '{key}' should be a dict."
+            validate_dict_structure(value, expected_type)
+
+        elif isinstance(expected_type, type):
+            assert isinstance(
+                value, expected_type
+            ), f"Key '{key}' should be of type {expected_type}, but got {type(value)}"
+
+        else:
+            raise ValueError(f"Unexpected type for key '{key}': {expected_type}")
+
+
+def verify_uuid(uuid_to_test, version=4):
+    try:
+        uuid_obj = UUID(uuid_to_test, version=version)
+    except ValueError:
+        assert False, "Not a valid UUID"
+    assert str(uuid_obj) == uuid_to_test, "Not a valid UUID"
+
+
+async def send_http_request(
+    connection, endpoint, method, data=None, expect_response=True
+):
+    curl_command = [
+        "curl",
+        "-X",
+        method,
+        endpoint,
+        "-H",
+        "Content-Type: application/json",
+    ]
+
+    if data:
+        curl_command.extend(["-d", data])
+
+    process = await connection.create_process(curl_command).execute()
+    response = process.get_stdout()
+    if expect_response:
+        try:
+            return json.loads(response)
+        except json.JSONDecodeError:
+            assert False, f"Expected JSON response but got: {response}"
+    return None
+
+
+async def clean_up_machines(connection, server_host):
+    machines = await send_http_request(
+        connection, f"{server_host}/v1/meshnet/machines", method="GET"
+    )
+
+    for machine in machines:
+        await send_http_request(
+            connection,
+            f"{server_host}/v1/meshnet/machines/{machine['identifier']}",
+            method="DELETE",
+            expect_response=False,
+        )
+
+
+linux_vm = MachineData(
+    public_key="zIdfzkxAhNKLj5ayl232xO/cnFm+v/9BK47mOdAxB1c=",
+    hardware_identifier="01c191a655e30be58dbe48c6a1c1508a",
```

Comment:
This still raises me a question. Maybe it makes sense though but as addressed in the previous comment:
> Well, we don't have any validation on keys. From API perspective string type is expected.

If these are not validated in any way but just need to be strings, I think it might be beneficial to use obvious strings, something like `"linux_vm_public_key_str"`.

My main motivation is to reduce complexity, seeing real-looking public key looks like it's generated somehow or must correlate with something :)

Reply:
It make sense. I think I misunderstood your previous comment.